README.md
XBMC LCDproc Python addon for Futaba mdm166a.
=============================================

Copyright (C) 2012 Team XBMC, Daniel 'herrnst' Scheller
Based on initial work (C) 2012 Memphiz/Team XBMC
/modded by D.K.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

See LICENSE.txt for more information.


changelog.txt
  [B]1.0.2[/B]
  
- Changed dimming during pause.
  
  [B]1.0.1[/B]
  
- Update for Kodi 14.x Helix
  
  [B]0.0.1[/B]

First version for testing based on script.xbmc.lcdproc v2.6.0 for Kodi 14.

- return to [B]Gotham 13.2[/B] support
- add @ icon for internet stream for mdm166a
- add separate screen for Music Videos
- add Polish translations
- change config file to [B]VFD.xml[/B]
- use mutebug infobool to determine mute state

lcdmain.py
'''
    XBMC LCDproc addon
    Copyright (C) 2012 Team XBMC
    Copyright (C) 2012 Daniel 'herrnst' Scheller
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import xbmc
import xbmcaddon
import time
import os

__settings__   = xbmcaddon.Addon(id='script.mdm166a')
__cwd__        = __settings__.getAddonInfo('path')
__icon__       = os.path.join(__cwd__,"icon.png")
__scriptname__ = "Futaba mdm166a"

BASE_RESOURCE_PATH = xbmc.translatePath( os.path.join( __cwd__, 'resources', 'lib' ) )
sys.path.insert(0, BASE_RESOURCE_PATH)

from settings import *
from lcdproc import *
from infolabels import *

global g_failedConnectionNotified
global g_initialConnectAttempt
global g_lcdproc

def initGlobals():
  global g_failedConnectionNotified
  global g_initialConnectAttempt
  global g_lcdproc 

  g_failedConnectionNotified = False   
  g_initialConnectAttempt = True
  settings_initGlobals()
  g_lcdproc = LCDProc()

  InfoLabel_Initialize()

# handle dispay of connection notificaiton popups
def HandleConnectionNotification(bConnectSuccess):
  global g_failedConnectionNotified
  global g_initialConnectAttempt

  if not bConnectSuccess:
    if not g_failedConnectionNotified:
      g_failedConnectionNotified = True
      g_initialConnectAttempt = False
      text = __settings__.getLocalizedString(32500)
      xbmc.executebuiltin("XBMC.Notification(%s,%s,%s,%s)" % (__scriptname__,text,10,__icon__))
  else:
    text = __settings__.getLocalizedString(32501)
    if not g_initialConnectAttempt:
      xbmc.executebuiltin("XBMC.Notification(%s,%s,%s,%s)" % (__scriptname__,text,10,__icon__))
      g_failedConnectionNotified = True

# returns mode identifier based on currently playing media/active navigation
def getLcdMode():                 
  ret = LCD_MODE.LCD_MODE_GENERAL

  navActive = InfoLabel_IsNavigationActive()
  screenSaver = InfoLabel_IsScreenSaverActive()
  playingVideo = InfoLabel_PlayingVideo()
  playingTVShow = InfoLabel_PlayingTVShow()
  playingMusicvideo = InfoLabel_PlayingMusicvideo()
  playingMusic = InfoLabel_PlayingAudio()
  playingPVRTV = InfoLabel_PlayingLiveTV()
  playingPVRRadio = InfoLabel_PlayingLiveRadio()

  if navActive:
    ret = LCD_MODE.LCD_MODE_NAVIGATION
  elif screenSaver:
    ret = LCD_MODE.LCD_MODE_SCREENSAVER
  elif playingPVRTV:
    ret = LCD_MODE.LCD_MODE_PVRTV
  elif playingPVRRadio:
    ret = LCD_MODE.LCD_MODE_PVRRADIO
  elif playingTVShow:
    ret = LCD_MODE.LCD_MODE_TVSHOW
  elif playingMusicvideo:
    ret = LCD_MODE.LCD_MODE_MUSICVIDEO
  elif playingVideo:
    ret = LCD_MODE.LCD_MODE_VIDEO
  elif playingMusic:
    ret = LCD_MODE.LCD_MODE_MUSIC
   
  return ret

def process_lcd():

  while not xbmc.abortRequested:
    if handleConnectLCD():
      settingsChanged = settings_didSettingsChange()

      if settingsChanged:
        g_lcdproc.UpdateGUISettings()

      g_lcdproc.Render(getLcdMode(), settingsChanged)

    time.sleep(1.0 / float(settings_getRefreshRate())) # refresh after configured rate

  g_lcdproc.Shutdown()

def handleConnectLCD():
  ret = True

  # make sure not to block things when shutdown is requested
  if not xbmc.abortRequested:
    #check for new settings
    if settings_checkForNewSettings() or not g_lcdproc.IsConnected():    #networksettings changed?
      g_failedConnectionNotified = False  #reset notification flag

      ret = g_lcdproc.Initialize()
      if not settings_getHideConnPopups():
        HandleConnectionNotification(ret)

  return ret

#########################################

#MAIN - entry point

# init vars and classes
initGlobals()

# initialise and load GUI settings
settings_setup()

# do LCD processing loop (needs to catch xbmc.abortRequested !)
process_lcd()


resources/lib/extraicons.py
'''
    XBMC LCDproc addon
    Copyright (C) 2012 Team XBMC
    
    Extra icon defines/enums
    Copyright (C) 2012 Daniel 'herrnst' Scheller
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

# enum snippet from http://stackoverflow.com/a/1695250 - thanks!
def enum(*sequential, **named):
  enums = dict(zip(sequential, range(len(sequential))), **named)
  return type('Enum', (), enums)

LCD_EXTRABARS_MAX = 4

LCD_EXTRAICONS = enum(
  'LCD_EXTRAICON_NONE',
  'LCD_EXTRAICON_PLAYING',
  'LCD_EXTRAICON_PAUSE',
  'LCD_EXTRAICON_MOVIE',
  'LCD_EXTRAICON_MUSIC',
  'LCD_EXTRAICON_WEATHER',
  'LCD_EXTRAICON_TV',
  'LCD_EXTRAICON_PHOTO',
  'LCD_EXTRAICON_WEBCASTING',
  'LCD_EXTRAICON_MUTE',
  'LCD_EXTRAICON_AT',
  'LCD_EXTRAICON_REPEAT',
  'LCD_EXTRAICON_SHUFFLE',
  'LCD_EXTRAICON_ALARM',
  'LCD_EXTRAICON_RECORD',
  'LCD_EXTRAICON_VOLUME',
  'LCD_EXTRAICON_TIME',
  'LCD_EXTRAICON_SPDIF',
  'LCD_EXTRAICON_DISC_IN',
  'LCD_EXTRAICON_SCR1',
  'LCD_EXTRAICON_SCR2',
  'LCD_EXTRAICON_RESOLUTION_SD',
  'LCD_EXTRAICON_RESOLUTION_HD',
  'LCD_EXTRAICON_VCODEC_MPEG',
  'LCD_EXTRAICON_VCODEC_DIVX',
  'LCD_EXTRAICON_VCODEC_XVID',
  'LCD_EXTRAICON_VCODEC_WMV',
  'LCD_EXTRAICON_ACODEC_MPEG',
  'LCD_EXTRAICON_ACODEC_AC3',
  'LCD_EXTRAICON_ACODEC_DTS',
  'LCD_EXTRAICON_ACODEC_VWMA', # e.g. iMON has video-WMA AND audio-WMA...
  'LCD_EXTRAICON_ACODEC_MP3',
  'LCD_EXTRAICON_ACODEC_OGG',
  'LCD_EXTRAICON_ACODEC_AWMA', # see ACODEC_VWMA
  'LCD_EXTRAICON_ACODEC_WAV',
  'LCD_EXTRAICON_OUTSOURCE',
  'LCD_EXTRAICON_OUTFIT',
  'LCD_EXTRAICON_OUT_2_0',
  'LCD_EXTRAICON_OUT_5_1',
  'LCD_EXTRAICON_OUT_7_1',
  'LCD_EXTRAICON_MAX'
)

LCD_EXTRAICONCATEGORIES = enum(
  'LCD_ICONCAT_MODES',
  'LCD_ICONCAT_OUTSCALE',
  'LCD_ICONCAT_CODECS',
  'LCD_ICONCAT_VIDEOCODECS',
  'LCD_ICONCAT_AUDIOCODECS',
  'LCD_ICONCAT_AUDIOCHANNELS'
)


resources/lib/infolabels.py
'''
    XBMC LCDproc addon
    Copyright (C) 2012 Team XBMC
    
    InfoLabel handling
    Copyright (C) 2012 Daniel 'herrnst' Scheller
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import os
import string
import sys
import time

import xbmc
import xbmcaddon
import xbmcgui

# wonder why "from settings import *" does not work...
__settingshandler__ = sys.modules[ "settings" ]

# enum snippet from http://stackoverflow.com/a/1695250 - thanks!
def enum(*sequential, **named):
  enums = dict(zip(sequential, range(len(sequential))), **named)
  return type('Enum', (), enums)

# interesting XBMC GUI Window IDs (no defines seem to exist for this)
class WINDOW_IDS:
  WINDOW_WEATHER               = 12600
  WINDOW_PVR                   = 10601
  WINDOW_PVR_MAX               = 10699
  WINDOW_VIDEOS                = 10006
  WINDOW_VIDEO_FILES           = 10024
  WINDOW_VIDEO_NAV             = 10025
  WINDOW_VIDEO_PLAYLIST        = 10028
  WINDOW_MUSIC                 = 10005
  WINDOW_MUSIC_PLAYLIST        = 10500
  WINDOW_MUSIC_FILES           = 10501
  WINDOW_MUSIC_NAV             = 10502
  WINDOW_MUSIC_PLAYLIST_EDITOR = 10503
  WINDOW_PICTURES              = 10002
  WINDOW_DIALOG_VOLUME_BAR     = 10104
  WINDOW_DIALOG_KAI_TOAST      = 10107
  
g_StreamPrefixes = [ \
 "http", "https", "tcp",   "udp",    "rtp",  \
 "sdp",  "mms",   "mmst",  "mmsh",   "rtsp", \
 "rtmp", "rtmpt", "rtmpe", "rtmpte", "rtmps" \
]

global g_InfoLabel_oldMenu
global g_InfoLabel_oldSubMenu
global g_InfoLabel_navTimer
global g_InfoLabel_oldFilenameandpath
global g_InfoLabel_CachedFilenameIsStream

def InfoLabel_Initialize():
  global g_InfoLabel_oldMenu
  global g_InfoLabel_oldSubMenu
  global g_InfoLabel_navTimer
  global g_InfoLabel_oldFilenameandpath
  global g_InfoLabel_CachedFilenameIsStream

  g_InfoLabel_oldMenu = ""
  g_InfoLabel_oldSubMenu = ""
  g_InfoLabel_navTimer = time.time()
  g_InfoLabel_oldFilenameandpath = ""
  g_InfoLabel_CachedFilenameIsStream = False

def InfoLabel_GetInfoLabel(strLabel):
  return xbmc.getInfoLabel(strLabel)

def InfoLabel_GetBool(strBool):
  return xbmc.getCondVisibility(strBool)

def InfoLabel_GetActiveWindowID():
  return int(xbmcgui.getCurrentWindowId())

def InfoLabel_timeToSecs(timeAr):
  # initialise return
  currentSecs = 0

  arLen = len(timeAr)
  if arLen == 1:
    currentSecs = int(timeAr[0])
  elif arLen == 2:
    currentSecs = int(timeAr[0]) * 60 + int(timeAr[1])
  elif arLen == 3:
    currentSecs = int(timeAr[0]) * 60 * 60 + int(timeAr[1]) * 60 + int(timeAr[2])

  return currentSecs

def InfoLabel_WindowIsActive(WindowID):
  return InfoLabel_GetBool("Window.IsActive(" + str(WindowID) + ")")

def InfoLabel_PlayingVideo():
  return InfoLabel_GetBool("Player.HasVideo")

def InfoLabel_PlayingTVShow():
  if InfoLabel_PlayingVideo() and len(InfoLabel_GetInfoLabel("VideoPlayer.TVShowTitle")): 
    return True
  else:
    return False

def InfoLabel_PlayingMusicvideo():
  return InfoLabel_GetBool("VideoPlayer.Content(musicvideos)")

def InfoLabel_PlayingAudio():
  return InfoLabel_GetBool("Player.HasAudio")

def InfoLabel_PlayingLiveTV():
  return InfoLabel_GetBool("PVR.IsPlayingTV")

def InfoLabel_PlayingLiveRadio():
  return InfoLabel_GetBool("PVR.IsPlayingRadio")

def InfoLabel_GetSystemTime():
  # apply some split magic for 12h format here, as "hh:mm:ss"
  # makes up for format guessing inside XBMC - fix for post-frodo at
  # https://github.com/xbmc/xbmc/pull/2321
  ret = "0" + InfoLabel_GetInfoLabel("System.Time(hh:mm:ss)").split(" ")[0]
  return ret[-8:]

def InfoLabel_GetPlayerTime():
  return InfoLabel_GetInfoLabel("Player.Time")

def InfoLabel_GetPlayerDuration():
  return InfoLabel_GetInfoLabel("Player.Duration")

def InfoLabel_IsPlayerPlaying():
  return InfoLabel_GetBool("Player.Playing")

def InfoLabel_IsPlayerPaused():
  return InfoLabel_GetBool("Player.Paused")

def InfoLabel_IsPlayerForwarding():
  return InfoLabel_GetBool("Player.Forwarding")

def InfoLabel_IsPlayerRewinding():
  return InfoLabel_GetBool("Player.Rewinding")

def InfoLabel_IsPlayingAny():
  return (InfoLabel_IsPlayerPlaying() |
          InfoLabel_IsPlayerPaused() |
          InfoLabel_IsPlayerForwarding() |
          InfoLabel_IsPlayerRewinding())

def InfoLabel_IsInternetStream():
  global g_InfoLabel_oldFilenameandpath
  global g_InfoLabel_CachedFilenameIsStream

  fname = InfoLabel_GetInfoLabel("Player.Filenameandpath")

  if fname != g_InfoLabel_oldFilenameandpath:
    g_InfoLabel_oldFilenameandpath = fname
    g_InfoLabel_CachedFilenameIsStream = False

    for prefix in g_StreamPrefixes:
      if fname.find(prefix + "://") == 0:
        g_InfoLabel_CachedFilenameIsStream = True

  return g_InfoLabel_CachedFilenameIsStream

def InfoLabel_IsPassthroughAudio():
  return InfoLabel_GetBool("Player.Passthrough")

def InfoLabel_IsPVRRecording():
  return InfoLabel_GetBool("PVR.IsRecording")

def InfoLabel_IsPlaylistRandom():
  return InfoLabel_GetBool("Playlist.IsRandom")

def InfoLabel_IsPlaylistRepeatAll():
  return InfoLabel_GetBool("Playlist.IsRepeat")

def InfoLabel_IsPlaylistRepeatOne():
  return InfoLabel_GetBool("Playlist.IsRepeatOne")

def InfoLabel_IsPlaylistRepeatAny():
  return (InfoLabel_IsPlaylistRepeatAll() | InfoLabel_IsPlaylistRepeatOne())

def InfoLabel_IsDiscInDrive():
  return InfoLabel_GetBool("System.HasMediaDVD")

def InfoLabel_IsScreenSaverActive():
  return InfoLabel_GetBool("System.ScreenSaverActive")

def InfoLabel_IsMuted():
  return InfoLabel_GetBool("Window.IsActive(mutebug)")

def InfoLabel_IsStreamAT():
  return InfoLabel_GetBool("Player.IsInternetStream")

def InfoLabel_GetVolumePercent():
  volumedb = float(string.replace(string.replace(InfoLabel_GetInfoLabel("Player.Volume"), ",", "."), " dB", ""))
  return (100 * (60.0 + volumedb) / 60)

def InfoLabel_GetPlayerTimeSecs():
  currentTimeAr = InfoLabel_GetPlayerTime().split(":")
  if currentTimeAr[0] == "":
    return 0

  return InfoLabel_timeToSecs(currentTimeAr)

def InfoLabel_GetPlayerDurationSecs():
  currentDurationAr = InfoLabel_GetPlayerDuration().split(":")
  if currentDurationAr[0] == "":
    return 0

  return InfoLabel_timeToSecs(currentDurationAr)

def InfoLabel_GetProgressPercent():
  tCurrent = InfoLabel_GetPlayerTimeSecs()
  tTotal = InfoLabel_GetPlayerDurationSecs()

  if float(tTotal) == 0.0:
    return 0

  return float(tCurrent)/float(tTotal)

def InfoLabel_IsNavigationActive():
  global g_InfoLabel_oldMenu
  global g_InfoLabel_oldSubMenu
  global g_InfoLabel_navTimer

  #from settings import settings_getNavTimeout

  ret = False

  navtimeout = __settingshandler__.settings_getNavTimeout()
  menu = InfoLabel_GetInfoLabel("$INFO[System.CurrentWindow]")
  subMenu = InfoLabel_GetInfoLabel("$INFO[System.CurrentControl]")

  if menu != g_InfoLabel_oldMenu or subMenu != g_InfoLabel_oldSubMenu or (g_InfoLabel_navTimer + navtimeout) > time.time():
    ret = True
    if menu != g_InfoLabel_oldMenu or subMenu != g_InfoLabel_oldSubMenu:
      g_InfoLabel_navTimer = time.time()      
    g_InfoLabel_oldMenu = menu
    g_InfoLabel_oldSubMenu = subMenu

  return ret

def InfoLabel_IsWindowIDPVR(iWindowID):
  if iWindowID >= WINDOW_IDS.WINDOW_PVR and iWindowID <= WINDOW_IDS.WINDOW_PVR_MAX:
    return True

  return False

def InfoLabel_IsWindowIDVideo(iWindowID):
  if iWindowID in [WINDOW_IDS.WINDOW_VIDEOS, WINDOW_IDS.WINDOW_VIDEO_FILES,
                   WINDOW_IDS.WINDOW_VIDEO_NAV, WINDOW_IDS.WINDOW_VIDEO_PLAYLIST]:
    return True

  return False

def InfoLabel_IsWindowIDMusic(iWindowID):
  if iWindowID in [WINDOW_IDS.WINDOW_MUSIC, WINDOW_IDS.WINDOW_MUSIC_PLAYLIST,
                   WINDOW_IDS.WINDOW_MUSIC_FILES, WINDOW_IDS.WINDOW_MUSIC_NAV,
                   WINDOW_IDS.WINDOW_MUSIC_PLAYLIST_EDITOR]:
    return True

  return False

def InfoLabel_IsWindowIDPictures(iWindowID):
  if iWindowID == WINDOW_IDS.WINDOW_PICTURES:
    return True

  return False

def InfoLabel_IsWindowIDWeather(iWindowID):
  if iWindowID == WINDOW_IDS.WINDOW_WEATHER:
    return True

  return False


resources/lib/lcdproc_extra_mdm166a.py
'''
    XBMC LCDproc addon
    Copyright (C) 2012 Team XBMC
    
    Support for extra symbols on Futaba/Targa USB mdm166a VFD displays
    Copyright (C) 2012 Daniel 'herrnst' Scheller
    Original C implementation (C) 2010 Christian Leuschen
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import xbmc
import sys

__scriptname__ = sys.modules[ "__main__" ].__scriptname__
__settings__ = sys.modules[ "__main__" ].__settings__
__cwd__ = sys.modules[ "__main__" ].__cwd__
__icon__ = sys.modules[ "__main__" ].__icon__

from lcdproc import *
from lcdbase import LCD_EXTRAICONS
from extraicons import *
from lcdproc_extra_base import *

def log(loglevel, msg):
  xbmc.log("### [%s] - %s" % (__scriptname__,msg,), level=loglevel) 
  
# extra icon bitmasks
class MDM166A_ICONS:
  ICON_PLAY             = 0x01 << 0
  ICON_PAUSE            = 0x01 << 1
  ICON_RECORD           = 0x01 << 2
  ICON_MESSAGE          = 0x01 << 3
  ICON_AT               = 0x01 << 4
  ICON_MUTE             = 0x01 << 5
  ICON_ANTENNA          = 0x01 << 6
  ICON_VOLUME           = 0x01 << 7
  ICON_ANTLOW           = 0x01 << 13
  ICON_ANTMED           = 0x01 << 14
  ICON_ANTHIGH          = (0x01 << 13) | (0x01 << 14)

  # clear masks
  ICON_CLEAR_ANTENNABAR = 0xffffffff &~ ((0x01 << 13) | (0x01 << 14))

class LCDproc_extra_mdm166a(LCDproc_extra_base):
  def __init__(self):
    self.m_iOutputValueOldIcons = 1
    self.m_iOutputValueIcons = 0

    LCDproc_extra_base.__init__(self)

  # private
  def _SetBarDo(self, barnum, percent):
    # progress bar
    if barnum == 1:
      bitmask = 0x003F8000
      bitshift = 15
      scale = 96
    # volume indicator
    elif barnum == 2:
      bitmask = 0x00001F00
      bitshift = 8
      scale = 28
    else:
      return

    if percent < 0:
      rpercent = 0
    elif percent > 100:
      rpercent = 100
    else:
      rpercent = percent

    self.m_iOutputValueIcons = (self.m_iOutputValueIcons &~ bitmask)
    self.m_iOutputValueIcons |= (int(scale * (rpercent / 100)) << bitshift) & bitmask

  # private
  def _SetIconStateDo(self, bitmask, state):
    if state:
      self.m_iOutputValueIcons |= bitmask
    else:
      self.m_iOutputValueIcons &= ~bitmask

  def Initialize(self):
    for i in range(1, 5):
      self.SetBar(i, float(0))

  def SetOutputIcons(self):
    ret = ""

    if self.m_iOutputValueIcons != self.m_iOutputValueOldIcons:
      self.m_iOutputValueOldIcons = self.m_iOutputValueIcons
      ret += "output %d\n" % (self.m_iOutputValueIcons)

    return ret

  def GetOutputCommands(self):
    return self.SetOutputIcons()

  def SetBar(self, barnum, percent):
    self._SetBarDo(barnum, percent)

  def SetIconState(self, icon, state):
    # General states
    if icon == LCD_EXTRAICONS.LCD_EXTRAICON_MUTE:
      self._SetIconStateDo(MDM166A_ICONS.ICON_MUTE, state)
      
    if icon == LCD_EXTRAICONS.LCD_EXTRAICON_AT:
      self._SetIconStateDo(MDM166A_ICONS.ICON_AT, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_PLAYING:
      self._SetIconStateDo(MDM166A_ICONS.ICON_PLAY, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_PAUSE:
      self._SetIconStateDo(MDM166A_ICONS.ICON_PAUSE, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_ALARM:
      self._SetIconStateDo(MDM166A_ICONS.ICON_MESSAGE, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_RECORD:
      self._SetIconStateDo(MDM166A_ICONS.ICON_RECORD, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_VOLUME:
      self._SetIconStateDo(MDM166A_ICONS.ICON_VOLUME, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_SPDIF:
      self._SetIconStateDo(MDM166A_ICONS.ICON_ANTENNA, state)

    # Output channels
    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_OUT_2_0:
      self.m_iOutputValueIcons &= MDM166A_ICONS.ICON_CLEAR_ANTENNABAR
      self._SetIconStateDo(MDM166A_ICONS.ICON_ANTLOW, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_OUT_5_1:
      self.m_iOutputValueIcons &= MDM166A_ICONS.ICON_CLEAR_ANTENNABAR
      self._SetIconStateDo(MDM166A_ICONS.ICON_ANTMED, state)

    elif icon == LCD_EXTRAICONS.LCD_EXTRAICON_OUT_7_1:
      self.m_iOutputValueIcons &= MDM166A_ICONS.ICON_CLEAR_ANTENNABAR
      self._SetIconStateDo(MDM166A_ICONS.ICON_ANTHIGH, state)

  def ClearIconStates(self, category):
    if category == LCD_EXTRAICONCATEGORIES.LCD_ICONCAT_CODECS:
      self.m_iOutputValueIcons &= MDM166A_ICONS.ICON_CLEAR_ANTENNABAR

    elif category == LCD_EXTRAICONCATEGORIES.LCD_ICONCAT_AUDIOCHANNELS:
      self.m_iOutputValueIcons &= MDM166A_ICONS.ICON_CLEAR_ANTENNABAR

  def GetClearAllCmd(self):
    self.m_iOutputValueOldIcons = 0
    self.m_iOutputValueIcons = 0

    return "output 0\n"
